jgrapht-core/src/main/java/org/jgrapht/generate/DirectedScaleFreeGraphGenerator.java

```diff
@@ -0,0 +1,428 @@
+                String
+                    .format(
+                        "deltaIn and deltaOut values of (%f, %f) are invalid", deltaIn, deltaOut));
+        }
+        if (targetEdges < 0 && targetNodes < 0) {
+            throw new IllegalArgumentException(
+                "can not have both targetEdges and targetNodes not set.");
+        }
+
+    }
+
+    /**
+     * Generates an instance of the {@link Graph}.
+     * 
+     * @param target the target graph
+     * @param resultMap not used by this generator, can be null
+     * 
+     * @throws TooManyFailuresException When the method fails {@link #maxFailures} times to add a
+     *         new edge to the growing graph.
+     */
+    @Override
+    public void generateGraph(Graph<V, E> target, Map<String, V> resultMap)
+    {
+        Set<V> newNodesSet = new HashSet<>();
+        Set<E> newEdgesSet = new HashSet<>();
+
+        if (targetEdges == 0 || (targetEdges < 0 && targetNodes == 0))
+            return;
+        V initV = target.addVertex();
+        newNodesSet.add(initV);
+
+        int failuresCounter = 0;
+        // grow network now, edge by edge. If the number of edges is unlocked, continue growing not
+        // only until targetNodes is reached, but until adding any more edges would add another
+        // node. i.e. allow more beta growth but not alpha nor gamma.
+        while (targetEdges >= 0 ? targetEdges > newEdgesSet.size()
+            : targetNodes >= newNodesSet.size())
+        {
+
+            if (failuresCounter >= maxFailures) {
+                throw new TooManyFailuresException(
+                    failuresCounter + " consecutive failures is more than maximum allowed number ("
+                        + maxFailures + ").");
+            }
+
+            V v, w, newV = null, newW = null;
+            E e;
+            float tributaries = rng.nextFloat();
+            if (tributaries <= alpha) {
+                // stop adding nodes if you will exceed the target
+                if (targetEdges < 0 && newNodesSet.size() == targetNodes)
+                    break;
+                v = newV = target.addVertex();
+                w = pickAVertex(target, newNodesSet, newEdgesSet, Direction.IN, deltaIn);
+            } else if (tributaries <= alphaPlusBeta) {
+                v = pickAVertex(target, newNodesSet, newEdgesSet, Direction.OUT, deltaOut);
+                w = pickAVertex(target, newNodesSet, newEdgesSet, Direction.IN, deltaIn);
+            } else {// gamma
+                    // stop adding nodes if you will exceed the target
+                if (targetEdges < 0 && newNodesSet.size() == targetNodes)
+                    break;
+                v = pickAVertex(target, newNodesSet, newEdgesSet, Direction.OUT, deltaOut);
+                w = newW = target.addVertex();
+            }
+
+            if (v == null || w == null) {
+                if (v == null) {
+                    target.removeVertex(newW);
```

Comment:
Any chance this can be implemented without adding and removing vertices in the target graph? If not ignore my comment.

Reply:
I delayed adding a vertex as possible, and cached the last added vertex, so that it can be used in the next attempt, in case the current one fails.
Does this sound appealing?